feat(conditions): Add JsonCondition for direct JSON evaluation

theref opened the pull request:
## Summary

This PR introduces **JsonCondition**, a new condition type that evaluates JSON data directly without requiring an HTTP request. This provides a simpler, more general alternative to `JsonApiCondition` for cases where JSON data is already available.

## Motivation

Currently, `JsonApiCondition` is designed for fetching JSON from HTTPS endpoints and then querying it. However, there are use cases where:
- JSON data is already available via Context
- Data comes from Sequential Condition

`JsonCondition` fills this gap by providing direct JSON evaluation with optional JSONPath queries.

## Changes

### Core Implementation
1. **New `JsonCondition` class** (`nucypher/policy/conditions/json/json.py`)
   - Accepts JSON as dict, list, primitive, or JSON string
   - Optional JSONPath query support
   - Context variable resolution
   - Simpler architecture (inherits from `Condition` directly, not `ExecutionCallCondition`)
   - Uses standard `AnyField` for serialization

2. **Shared query logic** (`nucypher/policy/conditions/json/utils.py`)
   - Extracted `query_json_data()` utility function
   - Eliminates code duplication between `JsonCondition` and `JsonApiCondition`
   - Both condition types now use the same JSONPath query logic

3. **Type system integration**
   - Added `ConditionType.JSON` enum value
   - Added `JsonConditionDict` type definition
   - Registered in `ConditionLingo.resolve_condition_class()`

### Testing
- 11 comprehensive unit tests covering all functionality
- Tests include: initialization, validation, query execution, context variables, serialization, edge cases
- All existing tests still pass (351/351 condition tests passing)

## Usage Example

```python
from nucypher.policy.conditions.json.json import JsonCondition
from nucypher.policy.conditions.lingo import ReturnValueTest

# With dict and JSONPath query
condition = JsonCondition(
    data={"store": {"book": [{"price": 10.5}]}},
    query="$.store.book[0].price",
    return_value_test=ReturnValueTest("==", 10.5)
)

# With JSON string
condition = JsonCondition(
    data='{"value": 42}',
    query="$.value",
    return_value_test=ReturnValueTest("==", 42)
)

# With primitive (no query needed)
condition = JsonCondition(
    data=100,
    return_value_test=ReturnValueTest(">", 50)
)

# Verify
success, value = condition.verify()
```

## Review Notes

### Atomic Commits
The changes are organized into 4 logical commits:
1. **Refactor**: Extract shared JSON query logic
2. **Feature**: Add JsonCondition implementation
3. **Chore**: Register in type system
4. **Test**: Add comprehensive tests

Each commit is self-contained and can be reviewed independently.

### Code Quality
- ✅ All pre-commit hooks passing (ruff, darker, etc.)
- ✅ No regressions - all 351 existing condition tests pass
- ✅ Full test coverage for new functionality
- ✅ Follows existing patterns and conventions

### Backward Compatibility
- ✅ No breaking changes to existing APIs
- ✅ `JsonApiCondition` behavior unchanged (just uses shared utility)
- ✅ All existing tests pass without modification

## Related Issues

This implements a more basic, general JSON condition type as discussed in the context of making TACo conditions more flexible and composable.

## Checklist

- [x] Code follows project style guidelines
- [x] Tests added and passing
- [x] No breaking changes
- [x] Documentation updated (newsfragment included)
- [x] Pre-commit hooks passing

theref committed:
refactor(conditions): extract shared JSON query logic to utils

Extract JSONPath query logic from JsonRequestCall into a shared
query_json_data() utility function in json/utils.py. This eliminates
code duplication and enables reuse across different condition types.

Changes:
- Add query_json_data() function to json/utils.py
- Refactor JsonRequestCall._query_response() to use shared utility

theref committed:
feat(conditions): add JsonCondition for direct JSON evaluation

Add new JsonCondition class that evaluates JSON data directly without
requiring an HTTP request. Accepts JSON as dict, list, primitive, or
JSON string with optional JSONPath queries.

Features:
- Direct JSON input (no HTTP fetching)
- Optional JSONPath query support
- Context variable resolution
- Uses standard AnyField for serialization
- Simpler architecture (inherits from Condition directly)

theref committed:
chore(conditions): register JsonCondition in type system

Add JsonCondition to the condition type registry:
- Add ConditionType.JSON enum value
- Add JsonConditionDict type definition
- Register in ConditionLingo.resolve_condition_class()

theref committed:
test(conditions): add unit tests for JsonCondition

Add comprehensive test coverage for JsonCondition including:
- Initialization with various data types (dict, JSON string, primitives)
- Validation and error handling
- JSONPath query execution
- Context variable support
- Serialization/deserialization
- Edge cases (ambiguous queries, invalid JSON)

All 11 tests passing.

theref committed:
docs: add newsfragment for JsonCondition feature (#3656)

derekpierre commented on nucypher/policy/conditions/json/json.py:
```diff
@@ -0,0 +1,85 @@
+
+    try:
+        return json.loads(data)
+    except (json.JSONDecodeError, ValueError) as e:
```
`JSONDecodeError` is already a subclass of `ValueError`

```suggestion
    except (ValueError) as e:
```

derekpierre commented on nucypher/policy/conditions/json/json.py:
```diff
@@ -0,0 +1,85 @@
+        condition_type = fields.Str(
+            validate=validate.Equal(ConditionType.JSON.value), required=True
+        )
+        data = AnyField(required=True)
```
Would this simply always be a context variable? 🤔 

Analogous to `authorization_token` for `JsonApiCall` - https://github.com/nucypher/nucypher/blob/v7.7.x/nucypher/policy/conditions/json/api.py#L49

theref replied:
you're right, I can't think of any situation where this wouldn't be populated by context in some way. For some reason I don't want to enforce that though

derekpierre replied:
> For some reason I don't want to enforce that though

Could you expand on your apprehension? If not a context variable, it means that it could be hardcoded - is there a legit case for that?

theref replied:
i think it's just overly restrictive - what do we gain by restricting it to only context variables?

cygnusv replied:
As a reference, the `jwt_token` field in `JwtCondition` is required to be a context variable.

theref replied:
ok, didn't know that :) I don't feel overly strongly on this, so I'm happy to defer and follow you both

derekpierre replied:
> i think it's just overly restrictive - what do we gain by restricting it to only context variables?

One gain is avoiding backwards compatibility issues.

It's easier to go from allowing only context variable to allowing anything, vs allowing anything (and potential user misuse) and then trying to restrict to only context variables.

derekpierre commented on nucypher/policy/conditions/json/json.py:
```diff
@@ -0,0 +1,85 @@
+            validate=validate.Equal(ConditionType.JSON.value), required=True
+        )
+        data = AnyField(required=True)
+        query = JSONPathField(required=False, allow_none=True)
```
Should this be required? Is there a case for it not needing to be used?

It wasn't required for JSON Rest endpoints because those can return values and not JSON.

If you deem it required, then update the constructor type to not be Optional.

theref replied:
The JSON could also just return a value, I think there's a test for that...if there's not I'll add one

derekpierre replied:
Fair point. I guess in the case of it just forwarding along a JSON response it could just be value.

There's an prior test for JsonApiCondition for just a value - see https://github.com/nucypher/nucypher/blob/main/tests/unit/conditions/test_json_api_condition.py#L57

derekpierre commented on nucypher/policy/conditions/json/json.py:
```diff
@@ -0,0 +1,85 @@
+        resolved_return_value_test = self.return_value_test.with_resolved_context(
+            **context
+        )
+        eval_result = resolved_return_value_test.eval(result_for_eval)
```
🚀

derekpierre commented on nucypher/policy/conditions/types.py:
```diff
@@ -60,6 +60,11 @@ class ContractConditionDict(RPCConditionDict):
 
+class JsonConditionDict(BaseExecConditionDict):
+    data: Any
+    query: NotRequired[str]
```
Update if `query` deemed required.

derekpierre reviewed (commented):
Nice work - some questions/suggestions.

derekpierre commented:
Let's see if we can cover most/all of the missing 6 lines for code coverage in json/utils.py 😉 :

<img width="918" height="245" alt="Screenshot 2025-09-30 at 1 57 30 PM" src="https://github.com/user-attachments/assets/65c02393-b487-4311-9a4e-0f099d03ecde" />

(Assuming the coverage report is correct, I'm surprised those lines aren't covered by other tests).

manumonti commented on nucypher/policy/conditions/json/json.py:
```diff
@@ -0,0 +1,85 @@
+from nucypher.policy.conditions.lingo import AnyField, ConditionType, ReturnValueTest
+
+
+def _parse_json_data(data: Any) -> Any:
+    """
+    Parse JSON data if it's a string, otherwise return as-is.
+    """
```
I'm wondering if we could manage data as a string (not a JSON string). For example, data is an address `data = "0xc0ffee254729296a45a3885639AC7E10F9d54979"`, or `data = "enabled"`

theref replied:
those are both valid json, so would be handled fine

manumonti reviewed (approved):
Looks great to me. I have a suggestion though. 

BTW, I'm super fan of how this PR looks. Commits are clear and concise, with well-formed messages. The PR message adds all the necessary context and makes clear the purpose of the changes and how they were made. Awesome.

theref committed:
refactor(conditions): simplify exception handling in JsonCondition

Remove redundant json.JSONDecodeError from exception handler since it is
a subclass of ValueError. This addresses review feedback from PR #3656.

theref committed:
test(conditions): improve coverage for json/utils.py to 100%

Add 4 new test cases to cover error paths in query_json_data():
- Invalid JSONPath syntax caught during initialization
- No matches found for JSONPath query
- Context variable resolution with no matches
- Runtime parser errors via context variables

This achieves 100% code coverage for json/utils.py (up from ~92%),
addressing review feedback from PR #3656.

cygnusv commented on nucypher/policy/conditions/json/json.py:
```diff
@@ -0,0 +1,85 @@
+
+class JsonCondition(Condition):
+    """
+    A JSON condition evaluates data that is directly provided as JSON.
+    The data can be a dict, list, primitive value, or a JSON string.
```
I don't understand the motivation for this. Shouldn't we restrict this to strings?

theref replied:
it's so that we can handle sequential conditions, where the json comes in already parsed as an object

cygnusv replied:
Oh I see. Thanks for the explanation 👍

cygnusv commented on tests/unit/conditions/test_json_condition.py:
```diff
@@ -0,0 +1,194 @@
+    assert condition.data == {"value": 42}
+
+
+def test_json_condition_with_primitive():
+    condition = JsonCondition(
+        data=42,
+        return_value_test=ReturnValueTest("==", 42),
+    )
+    assert condition.data == 42
+    assert condition.query is None
```
Related to my comment above. I know that `"42"` is a valid JSON string, but `42` is just an integer.

theref replied:
I thnk `42` is also valid json, it's certainly accepted by a lot of json parsers, including the one we use.

cygnusv replied:
Right, my point was more about acceptable types (str vs integers), but since you mentioned before that with sequential conditions JSON data may be already parsed as objects, then I have no further comment 👍

theref committed:
feat(conditions): enforce context variable requirement for JsonCondition data

Require JsonCondition.data to be a context variable, following the same
pattern as JWTCondition.jwt_token and JsonApiCondition.authorization_token.
This ensures compatibility and allows for easier widening of requirements
in the future if needed.

Changes:
- Add @validates decorator to require data field be a context variable
- Auto-parse JSON strings from resolved context variables
- Update data type from Any to str in schema and type definitions
- Rewrite all tests to use context variables
- Add tests for JSON string parsing from context variables

Addresses review feedback from @derekpierre on PR #3656.

derekpierre reviewed (approved):
🎸
